### ISSUE TITLE
jnp.frexp() custom vjp

### ISSUE THREAD

**Issue by matthewdhoffman** [google/jax#4756]
It seems to me that we should be able to take gradients w.r.t. the mantissa output of `jnp.frexp`, but we can't:

```
def test(x):
  mantissa, exponent = jnp.frexp(x)
  return mantissa

x = 0.1234
# jax.grad:
print(grad(test)(x))  # 0.0
# Finite differences:
epsilon = 1e-5
print((test(x + epsilon) - test(x - epsilon)) / (2*epsilon))  # 8.001924
# A reasonable subgradient, I think:
print(2.**(-jnp.frexp(x)[1]))  # 8.0
```

Presumably `grad` gives 0 because `frexp` is implemented in terms of non-differentiable bit-shifting ops.

Thoughts? Is it a problem that we can only get a subgradient?

**Comment by matthewdhoffman** [comment]
PS: If you're wondering, I want this so I can play around with stochastic rounding, following @sharadmv's discussions with @jekbradbury.

**Comment by jakevdp** [comment]
This seems reasonable to me, but we'd have to think about how to handle the boundaries where the function is not differentiable (i.e. when `abs(x)` is a power of 2)

**Comment by matthewdhoffman** [comment]
Yeah, I misspoke when I said "subgradient" above, since the mantissa has discontinuities at powers of 2.

Is there a precedent for what to do with functions like this that are differentiable almost everywhere, but have undefined (or infinite) derivatives for a countable number of values?

**Comment by shoyer** [comment]
> Is there a precedent for what to do with functions like this that are differentiable almost everywhere, but have undefined (or infinite) derivatives for a countable number of values?

I think `max`/`relu` is probably the canonical example of this.

**Comment by matthewdhoffman** [comment]
> > Is there a precedent for what to do with functions like this that are differentiable almost everywhere, but have undefined (or infinite) derivatives for a countable number of values?
> 
> I think `max`/`relu` is probably the canonical example of this.

`max` and `relu` are continuous, though, so there's a well-defined subgradient. Here there's a discontinuity, so it's less clear (to me) what to return—`inf`? `nan`?

**Comment by shoyer** [comment]
For reference, here's a plot of the derivative calculated with finite differences:
![image](https://user-images.githubusercontent.com/1217238/98038505-19e97500-1dd2-11eb-85ce-f434d0e39730.png)

**Comment by matthewdhoffman** [comment]
Yes, it's fine almost everywhere but if you try to do finite differences exactly on a power of 2, you get nasty stuff:

```python
def test(x):
  mantissa, exponent = jnp.frexp(x)
  return mantissa

# Power of 2 instead of 0.1234.
x = 0.125
# Finite differences:
epsilon = 1e-5
print((test(x + epsilon) - test(x - epsilon)) / (2*epsilon))  # -24994.0
```
